make routing explicit & report unknown methods

feuGeneA opened the pull request:
Fixes #378 make routing explicit & report unknown methods

explicitly denote each method as being either routed locally, routed remotely, or unknown.  when unknown methods are requested, give errors back to clients.

- [x] re-factor rpc/router.go
- [X] clean up code for `make lint`
- [x] add godoc to publish known methods and routing table decisions
- [X] write client_test.go, likely needing mocks for go-ethereum/rpc and status-go/geth/params
- [x] investgate test failures (whisper_mailbox_test.go :118 and :212)
- [ ] test real clients of geth/rpc/client.go's CallContext, which is the client.go method I changed. (Those real clients are e2e/rpc, e2e/testing.go, geth/txqueue, call_raw.go's callSingleMethod)

Solves #378

feuGeneA commented:
Please let me know if I'm on the right track here.

Also, this is my first foray into using Go, so any advice on how to do things more idiomatically would be much appreciated.  (I just noticed that I've neglected to `make lint`, so getting through that should also teach me a thing or two about coding style.)

Thanks!

divan commented:
Hey @feuGeneA, thanks for PR, but I'm afraid that task requires a deeper understanding of Status internals and definitely a lot of questions to ask the core team. It basically about creating documentation describing the logic behind routing rules. I don't think routing code itself needs refactoring.

Anyway, here are few comments on the code that might be useful to you:
 - there are more than two routing options. Apart of "remote" and "local", we have own handlers implementations (like for `eth_accounts`), so those methods not routed neither to local geth nor to remote upstream, rather executed locally using handler functions.
 - uppercased const and types names are bad practice in Go and should not be used (I hope `go lint` says this, but anyway, it differs from, say, C++).
 - a code should be formatted using [go fmt](https://golang.org/cmd/gofmt/) or it's more advanced analog [goimports](https://godoc.org/golang.org/x/tools/cmd/goimports). It's highly recommended to configure your IDE or editor to execute formatting tool as a save hook, so your code is always formatted — it's crazy fast and it's a common approach for Go code

Resuming, I'm afraid this task is mostly about docs, not the code, and it's not a good fit for first contributions, I'm afraid. :/

feuGeneA commented:
Thank you for the feedback @divan.

I did indeed understand that documentation is the end goal.  But in my mind, stand-alone documentation gets stale as code gets updated and as a manual doc-update step will at some point get forgotten, even if we all do our best at due diligence.

So my approach was to re-factor route.go to be self-documenting, with route choices clearly laid out in code, and with documentation of that code auto-generated into [the rpc godoc page](https://godoc.org/github.com/status-im/status-go/geth/rpc).

Looking closer at our rpc godoc page, it looks like only the exported methods are documented there.  But I see that `man godoc` says "show documentation for all declarations, not just the exported ones."  I'm not sure of the context of that quote, but it suggests that we should be able to rig it up to provide an authoritative documentation of our routing choices, generated from the code that actually makes those choices.

I did see the use of Handlers in client.go.  And I wondered whether they were also another type of routing option.  It seems counterintuitive to me to have some routing decisions made inside a route.go but other routing decisions made outside of it.  Since you've confirmed it, I'm now wondering if we should either (a) move all the Handler stuff to route.go, so that that file fully encapsulates all routing decisions, or (b) get rid of route.go and move it's logic into client.go, so that all of the routing can at least be in one place.

In any case, what do you think about my idea of using godoc to make routing decisions be self-documenting?

feuGeneA commented:
@divan Can we use this PR as a forum to openly develop that understanding of Status internals, and to get those questions for the core team asked, answered, and acted upon?

I'm still feeling like it would be best to make routing decisions be coded in a way that can be documented by godoc.  But if you still feel this is purely a documentation task, then can you help point us at the resources needed to get that done?

As far as understanding Status internals, I'm trying to make some headway. Here is a diagram I had made of the Status internals that reach into route.go, either directly or indirectly.  It's what made me feel comfortable about my refactoring.  It is quite busy, but comprehensive.  (A disclaimer, as for all such diagrams: it is accurate at the time of this writing, but it will age poorly as code evolves.  Readers in the future: do not trust it.)  If we were to consider refactoring both routing and handlers, I would make another diagram centered on client.go rather than route.go.
![status-rpc-refactor-route](https://user-images.githubusercontent.com/7883777/33385780-765d9b1c-d4f7-11e7-90bf-640e07ef32f0.png)

feuGeneA commented:
@divan For the record, what made me feel compelled to refactor was your comment in #378: "This causes bugs that hard to detect with current tests."

In order to have tests to catch bugs like that in #377, I figured the code needs to change.  My thought was that all known methods should be accounted for in our routing, not just those destined for upstream; and that when an unaccounted-for method is encountered, we should raise a failure so that we know to go explicitly set the routing, rather than letting the router silently/implicitly fall through to an undesired destination, as happened in the #377 bug where "net_*" methods were silently going to the local node because they weren't in the list of remote methods.

Does that make any sense?  Or do we really want everything that's not remote to silently/implicitly fall back to local?  It seems difficult if not impossible to test for bugs like that in #377 when we're using such fall-back logic.

divan commented:
@feuGeneA it does make sense, but what I'm afraid is that this PR from "update routing docs" transforms into "refactor rpc package" :)

>  I'm now wondering if we should either (a) move all the Handler stuff to route.go so that that file fully encapsulates all routing decisions

a) seems like a good choice here. I would call it `routeInternally` for example, and `client.RegisterHandler` should register the handler in `router` type then.

godoc is always a good place to be updated and reviewed.

As for the place for clarifying routing rules, I think https://github.com/status-im/status-go/issues/378 is the best place. I created this issue mostly because I didn't have enough knowledge of the original decisions myself, and wanted to ping other core team devs. Basically, I see it as a list of simple rules for upstream/non-upstream configurations and then for each `internal` call (`eth_accounts`, `eth_sendTransaction`) a good description of why this method is implemented inside status.

divan commented:
Nice chart, btw. Did you make it manually or used some tool?

feuGeneA commented:
@divan Thanks again for the feedback.

With regards to your fear, you mean that #378 has morphed from doc to refactor, right?  This PR has been refactor from the start :)  Would you rather have a doc update right now so we can close #378, and then have a separate issue for getting routing decisions auto-documented via godoc?  Or are you happy enough to let #378 linger a bit longer while I go straight for the godoc solution?

In any case, I will take a stab at moving the Handler stuff into route.go, along with getting that simple list of rules (along with descriptions for internally handled ones) published via godoc.

The diagram I made semi-manually, using [Dia](https://wiki.gnome.org/Apps/Dia).  I manually created all of the package and object elements and the dependency links between them, and Dia handles routing all those dotted lines as I drag the elements around (which can take some fiddling to make it look right).  It took a not-insignificant amount of time, but it was an exercise that has immensely helped my understanding of what's going on.

divan commented:
@feuGeneA Yeah, I think it can be done as separate PR. But make sure to address those moments I've commented about uppercased const names etc.

feuGeneA committed:
unify upstream & handler routing of RPC methods, w/ routing table in godoc

moved all Handler code into router.go. the (*Client) methods callMethod(),
handlers(), RegisterHandler(), and setResultFromRPCResponse() were all moved
onto the (*router) struct (route.go), and all copied verbatim, except that
callMethod was renamed to callHandler.  (*Client).CallContext(), and an old
shell of (*Client).RegisterHandler(), now just delegate to unexported methods
on (*router).

feuGeneA committed:
alphabetized method list in RoutingDestinations

feuGeneA commented:
While there's still a little work left to do here (I updated the checklist in my initial comment), I would appreciate a review of at least [the publish method routing decisions via godoc commit](https://github.com/status-im/status-go/pull/481/commits/df13dddc93cb52ba7f6f05e64f57ff9b60dd00d6), just to (in)validate my approach here.  You can preview the new godoc page [here](http://54.175.239.79:6060/pkg/github.com/status-im/status-go/geth/rpc/).  Thanks!

feuGeneA committed:
sync method routing with Alex Ewetumo's doc

doc @ https://docs.google.com/spreadsheets/d/1N1nuzVN5tXoDmzkBLeC9_mwIlVH8DGF7YD2XwxA8BAE/edit#gid=0

BUT, didn't change routing of net_*, which per doc above should be Local, but
per change by @adambabik should be Upstream.  (See change at
https://github.com/status-im/status-go/commit/e61c39b0b23e68d5002fb9b01425a1e55182fbba
)

feuGeneA committed:
additional methods revealed by e2e tests

feuGeneA commented:
there is an outstanding test failure when running with this branch:
```
gene@precision5510:~/go/src/github.com/status-im/status-go$ go test -timeout=20m ./e2e/whisper/... -network=
--- FAIL: TestRequestMessageFromMailboxAsync (4.32s)
        Error Trace:    whisper_mailbox_test.go:118
	Error:      	Not equal:
	            	expected: 1
	            	received: 0
--- FAIL: TestRequestMessagesInGroupChat (7.03s)
        Error Trace:    whisper_mailbox_test.go:212
	Error:      	Not equal:
	            	expected: 1
	            	received: 0
--- FAIL: TestWhisperMailboxTestSuite (11.35s)
FAIL
FAIL	github.com/status-im/status-go/e2e/whisper	32.830s
```

feuGeneA committed:
reduce Cognitive Complexity for Code Climate

feuGeneA commented:
@adambabik submitted changes per our discussion.  I took a few liberties beyond your specific suggestions, where I thought it made sense.  Please don't hesitate to ask for more changes where those liberties were unwarranted.

adambabik commented on geth/rpc/client.go:
```diff
@@ -38,23 +25,19 @@ type Client struct {
 	}
 
-	c.router = newRouter(c.upstreamEnabled)
+	c.router = newRouter(client, upstreamNode, upstream.Enabled)
```
I am not really convinced why router needs to know about all these clients. It looks like tight coupling. `router` in this case should only provide a relation between methods and destination.

The previous implementation was good and here we basically want a `switch` statement based on router output:
```go
destination := c.router.route(method)

switch destination {
case LocalHandler:
  // throw err if handler does not exist
  // also has the same interface as CallContext
  return c.callHandler(ctx, result, method, args...) 
case UpstreamNode:
  if c.upstreamEnabled {
    return c.upstream.CallContext(ctx, result, method, args...)
  }
  fallthrough // used it for the first time :D
default:
  return c.local.CallContext(ctx, result, method, args...)
}
```

adambabik commented on geth/rpc/route.go:
```diff
@@ -1,92 +1,282 @@
+	return nil
+}
+
+const upstreamNode = "UpstreamNode"
```
You can group these lines like this:
```go
const (
  upstreamNode = "UpstreamNode"
  localNode = "LocalNode"
  localHandler = "LocalHandler"
)
```

adambabik commented on geth/rpc/route.go:
```diff
@@ -1,92 +1,282 @@
+//
+// shh_getFilterMessages added to this list because of its use in
+// e2e/whisper/whisper_mailbox_test.go.
+var RoutingTable = map[string]string{
```
Let's kee here only routing to `localHandler` and `upstreamNode`. The rest should be routed to the local node by default.

adambabik commented:
Closing due to inactivity. Please reopen if it's still valid and there are new changes.
